private/buf/buflsp/mutex.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file defines various concurrency helpers.
+
+package buflsp
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+)
+
+const poison = ^uint64(0)
+
+var nextRequestID atomic.Uint64
+
+// withReentrancy assigns a unique request ID to the given context, which can be retrieved
+// with with getRequestID.
+func withRequestID(ctx context.Context) context.Context {
+	// This will always be unique. It is impossible to increment a uint64 and wrap around before
+	// the heat death of the universe.
+	id := nextRequestID.Add(1)
+	// We need to give the context package a unique identifier for the request; it can be
+	// any value. The address of the global we mint new IDs from is actually great for this,
+	// because users can't access it outside of this package, nor can they extract it out
+	// of the context itself.
+	return context.WithValue(ctx, &nextRequestID, id)
+}
+
+// getRequestID returns the request ID for this context, or 0 if ctx is nil or has no
+// such ID.
+func getRequestID(ctx context.Context) uint64 {
+	if ctx == nil {
+		return 0
+	}
+	id, ok := ctx.Value(&nextRequestID).(uint64)
+	if !ok {
+		return 0
+	}
+
+	// Make sure we don't return 0. This is the only place where the id is actually
+	// witnessed so doing +1 won't affect anything.
+	return id + 1
+}
```

Comment:
These are both helpers, so they can be at the bottom of the file, even though `withRequestID` is shared by another file. Also, the docs should be `%s/withReentrancy/withRequestID`.